### PR TITLE
ovnkube support for ovs offload

### PR DESCRIPTION
**Pull request by moshe010**
The current OVS supported is OVS running in Linux Kernel. Mellanox Accelerated Switching And Packet Processing (ASAP2 ) technology allows OVS offloading by handling OVS data-plane in Mellanox ConnectX-4 Lx onwards NIC hardware (Mellanox Embedded Switch or eSwitch) while maintaining OVS control-plane unmodified. OVS hardware offload is using SR-IOV technology with VF representor host net-device. The VF representor plays the same role as TAP devices in Para-Virtual (PV) setup. A packet sent through the VF representor on the host arrives to the VF, and a packet sent through the VF is received by its representor.

This proposal provide a mechanism to enable OVS hardware offload with OVN Kubernetes. 

More information can be found on Design Proposal for OVS Offload Support for OVN CNI [1]

[1] - https://docs.google.com/document/d/11uBezjSZg5b2g6VNs2mugGrgK5ZX9c-Xirj6XL_acjM/edit?usp=sharing

Co-authored-by: mamduhala mamduhala@mellanox.com
Signed-off-by: Moshe Levi <moshele@mellanox.com>

### CODE REVIEW

**Comment by girishmg** [comment]
@moshe010 Can you please fix all the lint errors as shown here:  https://travis-ci.org/openvswitch/ovn-kubernetes/builds/514959604?utm_source=github_status&utm_medium=notification?

**Comment by iyervl** [comment]
@moshe010  does it make sense to move all the SRIOV specific functions to a lib (e.g sriovnet.go) . Nit: typo in "SetupNetwrok". netlink.LinkSetHardwareAddr should work, since it will update the VF MAC when the pod comes up, but is there a case to change the VF's Mac (i.e. unbind/bind after the change) persistently?

**Comment by adrianchiris** [comment]
Hi,
@moshe010 is on PTO for the coming week. the PR will be updated next week.

**Comment by moshe010** [comment]
@iyervl, Sorry for the late response .Yes I am working with parav to move all the SR-IOV function to sriovnet, Once I am don't I will update the PR

File: go-controller/pkg/cni/helper_linux.go [hunk 1]
```diff
@@ -54,48 +94,112 @@ func setupInterface(netns ns.NetNS, containerID, ifName, macAddress, ipAddress,
...
 
-	// rename the host end of veth pair
-	hostIface.Name = containerID[:15]
-	if err := renameLink(oldHostVethName, hostIface.Name); err != nil {
-		return nil, nil, fmt.Errorf("failed to rename %s to %s: %v", oldHostVethName, hostIface.Name, err)
+	// rename the VF representor
+	if err := renameLink(hostIface.Name, containerID[:15]); err != nil {
+		return nil, nil, fmt.Errorf("failed to rename %s to %s: %v", hostIface.Name, containerID[:15], err)
 	}
+	hostIface.Name = containerID[:15]
```
Comment (girishmg) [inline go-controller/pkg/cni/helper_linux.go]: L198-202 is already done at L152-156. So, please remove the duplicate code.

File: go-controller/pkg/cni/types/types.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package types
```
Comment (girishmg) [inline go-controller/pkg/cni/types/types.go]: Not sure why we need this new file. Can't you move the new Netconf definition to  `go-controller/pkg/cni/types.go` itself?

**Comment by moshe010** [comment]
> @moshe010 I have two questions.
> 
> 1. The DeviceID is provided to OVN CNI by the Multus correct? We read that value into the new NetConf structure?
Yes
> 2. The new vendor libraries that you had to bring in, is it because of Mellanox sriovnet library?
Yes

File: go-controller/pkg/cni/types/types.go (hunk 2 above)
Comment (moshe010) [reply go-controller/pkg/cni/types/types.go]: the problem is cycle import in go. so currently cni package is using confg package. if I would add it to the types in the cni package I will need to import that in the config package. so to work around it with minimal changes I added the cni/types package.

File: go-controller/pkg/cni/helper_linux.go (hunk 1 above)
Comment (moshe010) [reply go-controller/pkg/cni/helper_linux.go]: yes you are right, I will remove it.

File: go-controller/pkg/cni/types/types.go (hunk 2 above)
Comment (girishmg) [reply go-controller/pkg/cni/types/types.go]: makes sense. thanks for explaining.

**Comment by girishmg** [comment]
@dcbw @shettyg FYI, I am working with @moshe010 on this trying to address few things...

**Comment by moshe010** [comment]
@girishmg any update on this PR?

**Comment by girishmg** [comment]
@moshe010

I deployed your PR on my rig with CX5 cards, and I was able to create pods with VF attached and VF_Rep in the host. It worked fine. I found few issues that needs to be addressed.

1. The MTU of the VF is 1500. It should be what is defined for the OVN cluster, and by default it is 1400. The `mtu` argument to the function setupSriovInterface() is being ignored, so can you please make use of it and set the right MTU on the VF?

2. I created net-attach-def resource called `ovnkube` in `default` namespace. However, coredns pod in kube-system failed to create since multus couldn't find  `ovnkube`. I moved the net-attach-def resource to kube-system and the coredns pods came up fine. However, when I had to deploy my application pods then I had to provide the following annotation:
      `v1.multus-cni.io/default-network: kube-system/ovnkube`
So, the application pod spec authors now need to refer to a system namespace resource. Am I missing, something?

3. Also, if you could please create a document in the `docs` directory to capture the steps required (create net-attach-def crd, and then resource, and then annotating POD spec, and so on) to plug a POD with VF and VF_REP in host. You have it in your google doc, but having it in the project repo will be better.

**Comment by moshe010** [comment]
@girishmg, I just pushed fix for the MTU so you can try it out. worked on my local setup
regarding 2 I am not sure I understand the issue (I guess I missing some knowledge here), but do you want to raise it on the multus meeting this week? 
regarding 3 in which file should I added the doc? or you meant that I will create a new file like ovs_offload.md?

**Comment by girishmg** [comment]
> regarding 3 in which file should I added the doc? or you meant that I will create a new file like ovs_offload.md?

Yes, please.

**Comment by girishmg** [comment]
@moshe010 

I tried the MTU changes, and it works as expected. Thanks. 

This PR looks good to me. Please let me know once you have the ovs_offload.md file ready.

Can you also please elaborate the commit message? Now, it just as 'ovnkube support for ovs offload'. Perhaps move the PR description into the commit message? Also,  include 'Fixes https://github.com/ovn-org/ovn-kubernetes/issues/773` in your commit message.

**Comment by girishmg** [comment]
@moshe010  Thanks for adding the documentation on how to use this feature. Your PR looks good to me.

@shettyg @dcbw I am going to integrate this today, so please let me know if you have any comments.

**Pull request merged**